11_FFTBloom/app_resources/kernel_spectrum_normalize.hlsl

```diff
@@ -0,0 +1,124 @@
+			const uint32_t2 zeroCoord = uint32_t2(indexDFT, 0);
+			const scalar_t shift = indexDFT & 1 ? scalar_t(-1) : scalar_t(1);
+			zero = (zero * shift) / power;
+			vector<scalar_t, 2> zeroVector = { zero.real(), zero.imag() };
+			kernelChannels[channel][zeroCoord] = zeroVector;
+
+			// Store nyquist element
+			const uint32_t2 nyquistCoord = uint32_t2(indexDFT, IMAGE_SIDE_LENGTH / 2);
+			// IMAGE_SIDE_LENGTH / 2 is even, so indexDFT + IMAGE_SIDE_LENGTH / 2 is even iff indexDFT is even, which then means the shift factor stays the same
+			nyquist = (nyquist * shift) / power;
+			vector<scalar_t, 2> nyquistVector = { nyquist.real(), nyquist.imag() };
+			kernelChannels[channel][nyquistCoord] = nyquistVector;
+		}
+	}
+	// The other rows have easier rules: They have to reflect their values along the Nyquist row
+	// If you'll remember, however, the first axis FFT stored the lower half of the DFT, bit-reversed not as a `log2(FFTSize)` bit number but in the range of the lower half 
+	// (so as a `log2(FFTSize / 2) bit number)`. Which means that whenever we get an element at `x' = index`, `y' = gl_WorkGroupID().x` we must get the actual coordinates of 
+	// the element in the DFT with `x = F(x')` and `y = bitreverse(y')`
+	else
+	{
+		for (uint32_t localElementIndex = 0; localElementIndex < ELEMENTS_PER_THREAD; localElementIndex++)
+		{
+			const uint32_t index = _NBL_HLSL_WORKGROUP_SIZE_ * localElementIndex | workgroup::SubgroupContiguousIndex();
+			// Get the element at `x' = index`, `y' = gl_WorkGroupID().x`
+			complex_t<scalar_t> toStore = rowMajorAccessor.get(rowMajorOffset(index, glsl::gl_WorkGroupID().x));
+
+			// Number of bits needed to represent the range of half the DFT
+			NBL_CONSTEXPR uint32_t bits = uint32_t(mpl::log2<IMAGE_SIDE_LENGTH>::value - 1);
+			uint32_t x = workgroup::fft::FFTIndexingUtils<ELEMENTS_PER_THREAD, _NBL_HLSL_WORKGROUP_SIZE_>::getDFTIndex(index);
+			uint32_t y = glsl::bitfieldReverse<uint32_t>(glsl::gl_WorkGroupID().x) >> (32 - bits);
+
+			// Store the element 
+			const scalar_t shift = (x + y) & 1 ? scalar_t(-1) : scalar_t(1);
+			toStore = (toStore * shift) / power;
+			vector<scalar_t, 2> toStoreVector = { toStore.real(), toStore.imag() };
+			kernelChannels[channel][uint32_t2(x, y)] = toStoreVector;
+
+			// Store the element at the column mirrored about the Nyquist column (so x'' = mirror(x))
+			// https://en.wikipedia.org/wiki/Discrete_Fourier_transform#Conjugation_in_time
+			// Guess what? The above says the row index is also the mirror about Nyquist! Neat
+			x = workgroup::fft::FFTIndexingUtils<ELEMENTS_PER_THREAD, _NBL_HLSL_WORKGROUP_SIZE_>::getDFTMirrorIndex(x);
+			y = workgroup::fft::FFTIndexingUtils<ELEMENTS_PER_THREAD, _NBL_HLSL_WORKGROUP_SIZE_>::getDFTMirrorIndex(y);
+			const complex_t<scalar_t> conjugated = conj(toStore);
+			toStoreVector.x = conjugated.real();
+			toStoreVector.y = conjugated.imag();
+			kernelChannels[channel][uint32_t2(x, y)] = toStoreVector;
+		}
+	}
```

Comment:
all this should have been in the second pass output (write directly to image)

Reply:
Okay so this should be done as write directly to image after each channel's FFT, then normalization step afterwards is launched which just reads a texel, dividides it by the power factor, and writes right back

Reply:
yes, keep it simple